In [1]:
import pandas as pd


In [2]:
# load the iris data

file_path = "./NYC_Tickets_2017.csv.crdownload"

df = pd.read_csv(file_path)

In [3]:
df

Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make  \
0              PAS   7/10/2016               7              SUBN        TOYOT   
1              PAS    7/8/2016               7              SUBN        TOYOT   
2              PAS   8/23/2016               5              SUBN         FORD   
3              COM   6/14/2017              47              REFG        MITSU   
4              COM  11/21/2016              69              DELV        INTER   
...            ...         ...             ...               ...          ...   
1048570        PAS    6/9/2017              36              4DSD        CHEVR   
1048571        PAS  12/11/2016              71              4DSD        NISSA   
1048572        PAS   4/13/2017              38              4DSD        HONDA   
1048573        PAS    4/4/2017              21              SUBN        VOLVO   
1048574        PAS   2/24/2017              50              4DSD        HONDA   

        Issuing Agency  Violation Location  Violation Precinct  \
0                    V                 NaN                   0   
1                    V                 NaN                   0   
2                    V                 NaN                   0   
3                    T                14.0                  14   
4                    T                13.0                  13   
...                ...                 ...                 ...   
1048570              V                 NaN                   0   
1048571              T                81.0                  81   
1048572              T                60.0                  60   
1048573              T                83.0                  83   
1048574              T               112.0                 112   

         Issuer Precinct  Issuer Code Issuer Command Violation Time  \
0                      0            0            NaN          0143A   
1                      0            0            NaN          0400P   
2                      0            0            NaN          0233P   
3                     14       359594           T102          1120A   
4                     13       364832           T102          0555P   
...                  ...          ...            ...            ...   
1048570                0            0            NaN          0200P   
1048571               81       357358           T301          1001A   
1048572               60       362530           T302          0202P   
1048573               83       364015           T301          0907A   
1048574              112       363063           T402          0458P   

        Violation County           Street Name  Law Section Sub Division  \
0                     BX  ALLERTON AVE (W/B) @         1111            D   
1                     BX  ALLERTON AVE (W/B) @         1111            D   
2                     BX  SB WEBSTER AVE @ E 1         1111            C   
3                     NY               7th Ave          408           l2   
4                     NY               6th Ave          408           h1   
...                  ...                   ...          ...          ...   
1048570               QN  WB HILLSIDE AVE @ 23         1180            B   
1048571                K            Putnam Ave          408           j6   
1048572                K      Coney Island Ave          408           h1   
1048573                K            Putnam Ave          408           d1   
1048574                Q              102nd St          408           e5   

        Vehicle Color  Vehicle Year           Violation Description  
0                  GY          2001    FAILURE TO STOP AT RED LIGHT  
1                  GY          2001    FAILURE TO STOP AT RED LIGHT  
2                  BK          2004              BUS LANE VIOLATION  
3                  WH          2007           47-Double PKG-Midtown  
4               WHITE          2007   69-Failure to Disp Muni Recpt  
...               ...           ...                     

In [4]:
# Drop null rows
df = df.dropna()
df

Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make  \
3              COM   6/14/2017              47              REFG        MITSU   
4              COM  11/21/2016              69              DELV        INTER   
11             COM   9/27/2016              19              DELV        KENWO   
15             PAS    7/7/2016              71              4DSD        VOLKS   
17             COM   1/26/2017              64               VAN        INTER   
...            ...         ...             ...               ...          ...   
1048569        PAS   3/12/2017              14              SUBN        ME/BE   
1048571        PAS  12/11/2016              71              4DSD        NISSA   
1048572        PAS   4/13/2017              38              4DSD        HONDA   
1048573        PAS    4/4/2017              21              SUBN        VOLVO   
1048574        PAS   2/24/2017              50              4DSD        HONDA   

        Issuing Agency  Violation Location  Violation Precinct  \
3                    T                14.0                  14   
4                    T                13.0                  13   
11                   T                18.0                  18   
15                   T               120.0                 120   
17                   T                17.0                  17   
...                ...                 ...                 ...   
1048569              T               109.0                 109   
1048571              T                81.0                  81   
1048572              T                60.0                  60   
1048573              T                83.0                  83   
1048574              T               112.0                 112   

         Issuer Precinct  Issuer Code Issuer Command Violation Time  \
3                     14       359594           T102          1120A   
4                     13       364832           T102          0555P   
11                    18       357355           T400          0707A   
15                   120       356557           T303          0645P   
17                    17       363557           T102          0256P   
...                  ...          ...            ...            ...   
1048569              109       357347           T401          1046A   
1048571               81       357358           T301          1001A   
1048572               60       362530           T302          0202P   
1048573               83       364015           T301          0907A   
1048574              112       363063           T402          0458P   

        Violation County       Street Name  Law Section Sub Division  \
3                     NY           7th Ave          408           l2   
4                     NY           6th Ave          408           h1   
11                    NY           6th Ave          408           c3   
15                     R   N St Austins Pl          408           j6   
17                    NY         E 43rd St          408           C8   
...                  ...               ...          ...          ...   
1048569                Q           Main St          408            c   
1048571                K        Putnam Ave          408           j6   
1048572                K  Coney Island Ave          408           h1   
1048573                K        Putnam Ave          408           d1   
1048574                Q          102nd St          408           e5   

        Vehicle Color  Vehicle Year           Violation Description  
3                  WH          2007           47-Double PKG-Midtown  
4               WHITE          2007   69-Failure to Disp Muni Recpt  
11                 RD          2013          19-No Stand (bus stop)  
15                 GY          2012   71-Insp. Sticker Missing (NYS  
17              BROWN          2007   64-No STD Ex Con/DPL, D/S Dec  
...               ...           ...                             ...  
1048569         BLACK             

In [5]:
#list dataframe data types
df.dtypes

Plate Type                object
Issue Date                object
Violation Code             int64
Vehicle Body Type         object
Vehicle Make              object
Issuing Agency            object
Violation Location       float64
Violation Precinct         int64
Issuer Precinct            int64
Issuer Code                int64
Issuer Command            object
Violation Time            object
Violation County          object
Street Name               object
Law Section                int64
Sub Division              object
Vehicle Color             object
Vehicle Year               int64
Violation Description     object
dtype: object

In [6]:
# Determine the number of unique values in each column 
df.nunique()

Plate Type                 70
Issue Date                361
Violation Code             79
Vehicle Body Type          53
Vehicle Make              226
Issuing Agency              1
Violation Location         77
Violation Precinct         77
Issuer Precinct            77
Issuer Code              2109
Issuer Command             33
Violation Time           1440
Violation County            5
Street Name              5759
Law Section                 1
Sub Division               93
Vehicle Color              94
Vehicle Year               75
Violation Description      97
dtype: int64

Violation time has 1440 unique value and it's an object. If we use this, we'll need to convert to date.<br> 
Issue date is also an object and has 361 unique values <br>
-BM

In [7]:
# drop issuing agency
df.drop(columns = ['Issuing Agency', 'Law Section'], inplace= True)


/Users/Beatriz/opt/anaconda3/envs/PythonData/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


# Take a look at the following columns:

Plate Type                 70 counted 
Issue Date                361
Violation Code             79 Counted
Vehicle Body Type          53 Counted
Vehicle Make              226 Counted
Issuing Agency              1 Dropped
Violation Location         77
Violation Precinct         77
Issuer Precinct            77
Issuer Code              2109 Counted
Issuer Command             33
Violation Time           1440
Violation County            5
Street Name              5759
Law Section                 1 Dropped
Sub Division               93
Vehicle Color              94 Counted
Vehicle Year               75
Violation Description      97
    
    
    
    
    
    
   Order Counted 
   
    Vehicle Color              94 object
    Plate Type                 70 object
    Violation Code             79 int 
    Vehicle Body Type          53
    Vehicle Make              226



In [8]:
# Check value distribution for 'Vehicle Color'
color_count = df['Vehicle Color'].value_counts()
color_count

WH       125914
GY       114848
WHITE    103684
BK        98650
BLACK     50144
          ...  
BRT           1
PR/           1
LTR           1
MRB           1
BR/           1
Name: Vehicle Color, Length: 94, dtype: int64

> consider dropping values with 1 count <br>
> consolidate colors (replace value "BK" and "BL" with "Black" (NEED TO FIGURE OUT CODE FOR THIS)

In [16]:
# Check value distribution for 'Plate Type' 70 object
Plate_Type_count = df['Plate Type'].value_counts()
Plate_Type_count

PAS    536737
COM    151060
OMT     25594
OMS      7331
SRF      5669
        ...  
LUA         1
ORM         1
WUG         1
LTR         1
FPW         1
Name: Plate Type, Length: 70, dtype: int64

In [17]:
# Check value distribution for 'Violation Code' 79 int
Violation_Code_count = df['Violation Code'].value_counts()
Violation_Code_count


21    120491
38    104002
14     79724
37     58257
20     54826
       ...  
97         1
44         1
28         1
92         1
93         1
Name: Violation Code, Length: 79, dtype: int64

In [18]:
# how many only show up once? 
less_than_1_Violation_Code_count= list(Violation_Code_count[Violation_Code_count ==1].index)
less_than_1_Violation_Code_count

[97, 44, 28, 92, 93]

In [19]:
# Check value distribution for 'Vehicle Body Type' 53 Object
Vehicle_Body_Type_count = df['Vehicle Body Type'].value_counts()
Vehicle_Body_Type_count


SUBN    254462
4DSD    233036
VAN     117425
DELV     55783
2DSD     22248
PICK     21671
REFG      9136
TRAC      6132
UTIL      5169
TAXI      3437
CONV      3318
TRLR      2592
BUS       2542
MCY       2139
TK        1948
FLAT      1106
TR/C       817
DUMP       769
T/CR       741
SEDN       487
STAK       406
TR         339
SEMI       316
BOAT       243
TOW        232
SWT        150
TANK       103
LIM         80
BS          74
P/SH        57
S/SP        53
MCC         48
H/WH        43
W/DR        31
MOBL        28
SN/P        27
TRAV        25
CUST        24
LTRL        20
TR/E        18
CMIX        13
H/TR        10
POLE         9
R/RD         9
EMVR         8
RD/S         6
RPLC         3
FPM          2
WSR          2
HRSE         1
H/IN         1
LSV          1
RBM          1
Name: Vehicle Body Type, dtype: int64

In [21]:
# Check value distribution for 'Vehicle Make' 226 object

Vehicle_Make_count = df['Vehicle Make'].value_counts()
Vehicle_Make_count

FORD     94656
TOYOT    77530
HONDA    71045
NISSA    58689
CHEVR    52819
         ...  
PTRB         1
CECK         1
UMC          1
ANAK         1
UNCB         1
Name: Vehicle Make, Length: 226, dtype: int64

In [22]:
# see rare vehicles 
rare_vehicle = list(Vehicle_Make_count[Vehicle_Make_count ==1].index)
rare_vehicle

['CORB',
 'BUSI',
 'AM/G',
 'LR/R',
 'SCAN',
 'TR/T',
 'TURT',
 'CITC',
 'BUGAT',
 'CHEB',
 'BOUN',
 'AL/D',
 'PICK',
 'SIMCA',
 'OPELP',
 'ST/C',
 'YUKO',
 'WK/H',
 'WE/S',
 'YANH',
 'MBEN',
 'ROCK',
 'WH/G',
 'FE/C',
 'Z',
 'ISUA',
 'Thom',
 'HIND',
 'FLAI',
 'ADD',
 'ITNE',
 'SERR',
 'LEIS',
 'CABC',
 'VPG',
 'NEWF',
 'JAMB',
 'KIIA',
 'BGTI',
 'MIDA',
 'MREC',
 'ITNL',
 'CHAM',
 'YUGO',
 'BWN',
 'HINI',
 'AXAT',
 'STL',
 'SOUT',
 'UD/N',
 'PTRB',
 'CECK',
 'UMC',
 'ANAK',
 'UNCB']

In [ ]:
# code for Binning or replacing value names 

# # Determine which values to replace if counts are less than ...?
# replace_INCOME_AMT = list(INCOME_AMT_count[INCOME_AMT_count < 500].index)


# # Replace in dataframe
# for app in replace_INCOME_AMT:
#     application_df.INCOME_AMT = application_df.INCOME_AMT.replace(app,"5M+")
    
# # Check to make sure binning was successful
# application_df.INCOME_AMT.value_counts()

In [ ]:
#Find null values
# for column in df_shopping.columns:
#         print(f"Column {column} has {df_shopping[column].isnull().sum()} null values ")

In [ ]:
# Remove the CustomerID Column
# df_shopping.drop(columns = ['CustomerID'], inplace= True)
# df_shopping.head()

In [ ]:
# output_file_path = "./clean_parking_violations.csv"
# new_df.to_csv(output_file_path, index=False)